extensions/s3/src/main/java/io/deephaven/extensions/s3/S3SeekableChannelProvider.java

```diff
@@ -62,56 +63,26 @@ final class S3SeekableChannelProvider implements SeekableChannelsProvider {
     private final S3AsyncClient s3AsyncClient;
     private final S3Instructions s3Instructions;
 
-    private SoftReference<KeyedObjectHashMap<URI, FileSizeInfo>> uriToFileSizeSoftRef;
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<S3SeekableChannelProvider, SoftReference> FILE_SIZE_CACHE_REF_UPDATER =
+            AtomicReferenceFieldUpdater.newUpdater(S3SeekableChannelProvider.class, SoftReference.class,
+                    "fileSizeCacheRef");
 
-    private static final class FileSizeInfo {
-        private final URI uri;
-        private final long size;
-
-        FileSizeInfo(@NotNull final URI uri, final long size) {
-            this.uri = Require.neqNull(uri, "uri");
-            this.size = size;
-        }
-
-        @Override
-        public int hashCode() {
-            return Objects.hash(uri, size);
-        }
-
-        @Override
-        public boolean equals(final Object other) {
-            if (this == other) {
-                return true;
-            }
-            if (other == null || getClass() != other.getClass()) {
-                return false;
-            }
-            final FileSizeInfo that = (FileSizeInfo) other;
-            return size == that.size && uri.equals(that.uri);
-        }
-    }
-
-    private static final KeyedObjectKey<URI, FileSizeInfo> URI_MATCH_KEY = new KeyedObjectKey.Basic<>() {
-        @Override
-        public URI getKey(@NotNull final FileSizeInfo value) {
-            return value.uri;
-        }
-    };
+    private volatile SoftReference<Map<URI, FileSizeInfo>> fileSizeCacheRef;
 
     S3SeekableChannelProvider(@NotNull final S3Instructions s3Instructions) {
         this.s3AsyncClient = S3AsyncClientFactory.getAsyncClient(s3Instructions);
         this.s3Instructions = s3Instructions;
-        this.uriToFileSizeSoftRef = new SoftReference<>(new KeyedObjectHashMap<>(URI_MATCH_KEY));
     }
 
     @Override
     public SeekableByteChannel getReadChannel(@NotNull final SeekableChannelContext channelContext,
             @NotNull final URI uri) {
         final S3Uri s3Uri = s3AsyncClient.utilities().parseUri(uri);
         // context is unused here, will be set before reading from the channel
-        final KeyedObjectHashMap<URI, FileSizeInfo> uriToFileSize = uriToFileSizeSoftRef.get();
-        if (uriToFileSize != null && uriToFileSize.containsKey(uri)) {
-            return new S3SeekableByteChannel(s3Uri, uriToFileSize.get(uri).size);
+        final Map<URI, FileSizeInfo> fileSizeCache = fileSizeCacheRef.get();
```

Comment:
This can NPE, since we removed the initializer. We could put the initializer back and simplify my loop with the suggestion below. Otherwise you need to complicate this code.